test/test_meta.py

```diff
@@ -0,0 +1,846 @@
+                    f"originally (*{args}, **{kwargs})") from e
+
+        rs = func(*args, **kwargs)
+
+        # TODO: also handle cases where func raise an exception
+
+        # For now, only attempt if we managed to convert all tensor types
+        # (if any of them failed, we're in a mixed device situation and
+        # this isn't well supported)
+        if do_meta and hit > 0 and miss == 0:
+            try:
+                # suppress warnings
+                with warnings.catch_warnings():
+                    warnings.simplefilter("ignore")
+                    meta_rs = func(*meta_args, **meta_kwargs)
+            except Exception as e:
+                suppress = False
+                if isinstance(e, NotImplementedError):
+                    m = RE_NOT_IMPLEMENTED_MSG.search(e.args[0])
+                    if m and m.group(1) not in ("aten::_efficientzerotensor", "aten::view_as_real"):
+                        suppress = True
+                if not suppress:
+                    raise RuntimeError(f"""\
+failed to run: {func}(
+    *{meta_args},
+    **{meta_kwargs}
+  )""") from e
+            else:
+                def test_assert(cond, msg):
+                    if not cond:
+                        raise RuntimeError(f"""\
+meta disagrees with real impl:
+{func}(
+    *{meta_args},
+    **{meta_kwargs}
+) = {meta_r}
+{msg}
+""")
+                flat_meta_rs, _ = tree_flatten(meta_rs)
+                flat_rs, _ = tree_flatten(rs)
+                self.test_case.assertEqual(len(flat_meta_rs), len(flat_rs))
+                for i, meta_r, r in zip(range(len(flat_rs)), flat_meta_rs, flat_rs):
+                    if isinstance(r, torch.Tensor):
+                        test_assert(isinstance(meta_r, torch.Tensor), f"but real {i}th result is Tensor")
+                        test_assert(meta_r.dtype == r.dtype, f"but real dtype was {r.dtype}")
+                        test_assert(meta_r.shape == r.shape, f"but real shape was {r.shape}")
+                        test_assert(meta_r.stride() == r.stride(), f"but real stride was {r.stride()}")
+                        test_assert(
+                            meta_r.storage_offset() == r.storage_offset(),
+                            f"but real storage_offset was {r.storage_offset()}")
+                        test_assert(meta_r.requires_grad == r.requires_grad, f"but real requires_grad was {r.requires_grad}")
+                        test_assert(meta_r.is_conj() == r.is_conj(), f"but real is_conj was {r.is_conj()}")
+                        test_assert(meta_r.is_neg() == r.is_neg(), f"but real is_neg was {r.is_neg()}")
+
+        return rs
+
+class TestMeta(TestCase):
+    @unittest.skipIf(TEST_WITH_ASAN, "Skipped under ASAN")
+    @onlyNativeDeviceTypes
+    @skipIfCrossRef
+    @suppress_warnings
+    @ops(op_db)
+    def test_meta(self, device, dtype, op):
+        func = op.get_op()
+        if func in meta_exclude_set:
```

Comment:
I used to do this, but this makes the test suite run *a lot* longer, like 4x longer (because throwing exceptions is expensive). As a compromise, I can add an optional mode to do it this way, but not the default.

Reply:
Yeah I understand it takes longer. If you don't want to do it let's not bother, we can add a mode later if we really want to

Reply:
So, when I'm doing OpInfo tests only, it's tolerable. So I've turned on expected failures, and shuffled things around so that I can NOT do that for crossref testing.